code/game/machinery/doors/firedoor.dm

```diff
@@ -23,12 +23,86 @@
+	var/hazard = FALSE
 
 /obj/machinery/door/firedoor/Initialize()
 	. = ..()
-	CalculateAffectingAreas()
+	var/area/my_area = get_area(src)
+	LAZYADD(my_area.firedoors, src)
+
+	RegisterSignal(src, COMSIG_MERGER_ADDING, .proc/MergerAdding)
+	RegisterSignal(src, COMSIG_MERGER_REMOVING, .proc/MergerRemoving)
+
+	return INITIALIZE_HINT_LATELOAD
+
+/obj/machinery/door/firedoor/LateInitialize()
+	. = ..()
+	GetMergeGroup(merger_id, allowed_types = typecacheof(/obj/machinery/door/firedoor))
+
+/obj/machinery/door/firedoor/Destroy()
+	var/area/my_area = get_area(src)
+	LAZYREMOVE(my_area.firedoors,src)
+	return..()
+
+///////////////////////////////////////////////////////////////////
+// Merger handling
+
+/obj/machinery/door/firedoor/proc/MergerAdding(obj/machinery/door/firedoor/us, datum/merger/new_merger)
+	SIGNAL_HANDLER
+	if(new_merger.id != merger_id)
+		return
+	RegisterSignal(new_merger, COMSIG_MERGER_REFRESH_COMPLETE, .proc/MergerRefreshComplete)
+
+/obj/machinery/door/firedoor/proc/MergerRemoving(obj/machinery/door/firedoor/us, datum/merger/old_merger)
+	SIGNAL_HANDLER
+	if(old_merger.id != merger_id)
+		return
+	UnregisterSignal(old_merger, COMSIG_MERGER_REFRESH_COMPLETE)
+
+/// Handles the firelocks to register only one signal per group
+/obj/machinery/door/firedoor/proc/MergerRefreshComplete(datum/merger/merger, list/leaving_members, list/joining_members)
+	SIGNAL_HANDLER
+	var/turf/open/turf1 = get_step(src, dir)
+	var/turf/open/turf2 = get_step(src, turn(dir, 180))
+	if(merger.origin != src)
+		registered_turfs = FALSE
+		UnregisterSignal(turf1, COMSIG_TURF_EXPOSE)
+		UnregisterSignal(turf2, COMSIG_TURF_EXPOSE)
+		return
+
+	if(!registered_turfs)
+		RegisterSignal(turf1, COMSIG_TURF_EXPOSE, .proc/check_closing)
+		RegisterSignal(turf2, COMSIG_TURF_EXPOSE, .proc/check_closing)
+		registered_turfs = TRUE
+
+///Check if the temperature and pressure changes are enough to cause body damage
+/obj/machinery/door/firedoor/proc/check_closing(datum/source, datum/gas_mixture/environment, exposed_temperature)
+	SIGNAL_HANDLER
+	var/temperature_hazard = TRUE
+	if(exposed_temperature > BODYTEMP_COLD_DAMAGE_LIMIT && exposed_temperature < BODYTEMP_HEAT_DAMAGE_LIMIT)
+		temperature_hazard = FALSE
+
+	var/environment_pressure = environment.return_pressure()
+	var/pressure_hazard = TRUE
+	if(environment_pressure > HAZARD_LOW_PRESSURE && environment_pressure < HAZARD_HIGH_PRESSURE)
+		pressure_hazard = FALSE
+
+	var/datum/merger/merge_group = GetMergeGroup(merger_id, allowed_types = typecacheof(/obj/machinery/door/firedoor))
+	if(!temperature_hazard && !pressure_hazard)
+		for(var/obj/machinery/door/firedoor/considered_door as anything in merge_group.members)
+			considered_door.hazard = FALSE
+			INVOKE_ASYNC(considered_door, .proc/open)
+		return
+
+	for(var/obj/machinery/door/firedoor/considered_door as anything in merge_group.members)
+		considered_door.hazard = TRUE
+		INVOKE_ASYNC(considered_door, .proc/close)
```

Comment:
Is there a way of detecting if the merger group was closing doors in the first place? opening all firelocks just because a tile near one processed seems a bit off

Reply:
i don't know, if there is i can't find it

Reply:
Could you track that?

Reply:
Reposting from discord in an explanation of what I'm asking for/why

>so that proc is fired when a tile near a firelock is activated. if it's not hazardous it calls open on all the doors, and if it is it'll close them all 
>A: I think calling open on all firelocks each time a turf near them processes is just unreasonably expensive B it means that firelocks closed by players will inconsistently open semi randomly.
>
> What I'd like is for the merger datum to track if it's got any reason to be closed. you might need to do this on a tile by tile basis, because it'd also make tracking if all the doors should open sane